tree/tree/src/TTree.cxx

```diff
@@ -2789,7 +2789,7 @@ TFile* TTree::ChangeFile(TFile* file)
       if (newfile) newfile->Append(obj);
       file->Remove(obj);
    }
-   delete file;
+   file->Delete();
```

Comment:
"Surprisingly" this actually request the deletion/freeing of the keys' area from the file rather than calling C++ operator delete.  Try
```suggestion
   file->TObject::Delete();
```

Reply:
Yes indeed that calls `TFile::Delete` which calls `TDirectoryFile::Delete`  hence the request of key deletion .`TObject::Delete` is never called.

Reply:
:sob:

Reply:
(dangerous violation of https://en.wikipedia.org/wiki/Liskov_substitution_principle )

Reply:
agreed ... :(

Reply:
better now?